### PR TITLE
Fixes odo describe for non pushed components

### LINKED ISSUE #2288: odo describe should work even for not pushed components
```
▶ odo describe
 ✗  component nodejs-nodejs-ex-rgop not pushed to the OpenShift cluster, use `odo push` to deploy the component
```

odo describe should work even on a component that is not pushed in a similar way as for pushed components.

We will need new field in normal output. `State` that will indicated if the component is pushed or not (`Pushed` vs `Not Pushed`)
```
▶ odo describe
Component Name: nodejs-nodejs-ex-rgop
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858
URLs:
 · http://nodejs-nodejs-ex-rgop-8080-app-ododemo.e8ca.engint.openshiftapps.com exposed via 8080
State: Pushed
```
```
▶ odo describe
Component Name: nodejs-nodejs-ex-rgop
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858
URLs:
 · will be exposed via 8080
State: Not Pushed

The component is not present on the cluster. You can run `odo push` to create it on the cluster.
```

Not sure about the last message. What do you think?

JSON output already has the `State` filed for pushed components. 
```
▶ odo describe -o json
{
    "kind": "Component",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "name": "nodejs-nodejs-ex-rgop",
        "namespace": "ododemo",
        "creationTimestamp": null
    },
    "spec": {
        "app": "app",
        "type": "nodejs",
        "source": "file://./",
        "url": [
            "nodejs-nodejs-ex-rgop-8080"
        ],
        "env": [
            {
                "name": "DEBUG_PORT",
                "value": "5858"
            }
        ],
        "ports": [
            "8080/TCP"
        ]
    },
    "status": {
        "state": "Pushed"
    }
}
```
We just have to make sure that it works even for unpushed instead of erroring out
```
▶ odo describe -o json
{
    "kind": "Error",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "creationTimestamp": null
    },
    "message": "component nodejs-nodejs-ex-rgop not pushed to the OpenShift cluster, use `odo push` to deploy the component"
}
```

### PR DESCRIPTION
**Pull request by mik-dass**
**What kind of PR is this?**

/kind enhancement

**What does does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #2288 

**How to test changes / Special notes to the reviewer**:

run `odo describe` in non pushed odo component directory

### CODE REVIEW

**Comment by dharmit** [comment]
```sh
$ odo config view
COMPONENT SETTINGS
------------------------------------------------
PARAMETER         CURRENT_VALUE
Type              nodejs
Application       app
Project           nphmecdnzx
SourceType        local
Ref               
SourceLocation    ./
Ports             8080/TCP
Name              nodejs-nodejs-ex-kvna
MinMemory         
MaxMemory         
DebugPort         
Ignore            
MinCPU            
MaxCPU            

# before doing odo push
$ odo describe
Component Name: nodejs-nodejs-ex-kvna
Type: nodejs
Source: ./

# after doing odo push
$ odo describe
Component Name: nodejs-nodejs-ex-kvna
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858
```
* should `Environment Variables` be mentioned in `odo describe` output when the component is not pushed? If we fetch them from the running pod then no. But `odo describe` prints the output quite fast so I don't think we're fetching them from running pod ever. So, should we print them when doing `odo describe` without doing `odo push`?
* Output of `Source` is different for the two cases. Should we have consistency there?

**Comment by dharmit** [comment]
URL created for a pushed component doesn't show up in `odo describe` output. But when we delete the component without doing `odo push`, the URL shows up in `odo describe` output. Is this expected behaviour? Shouldn't the URL be described in the first case as well?
```sh
$ odo url create url1
 ✓  URL url1 created for component: nodejs-nodejs-ex-kvna

To create URL on the OpenShift Cluster, please use `odo push`

# component was already pushed to the cluster
# doesn't describe URLs
$ odo describe
Component Name: nodejs-nodejs-ex-kvna
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858

# deleted the component
$ odo delete -f
 ✓  Deleting component nodejs-nodejs-ex-kvna [144ms]
 ✓  Component nodejs-nodejs-ex-kvna from application app has been deleted

# now it does describe the URL. 
$ odo describe
Component Name: nodejs-nodejs-ex-kvna
Type: nodejs
Source: ./
URLs:
 · URL named url1 will be exposed via 8080/TCP
```

**Comment by dharmit** [comment]
Steps to reproduce:
1. `odo create nodejs`
2. `odo push`
3. `odo describe` (no problems)
4. `odo url create`
5. `odo push`
6. `odo describe` (panics)

```sh
$ git clone https://github.com/sclorg/nodejs-ex .
fatal: destination path '.' already exists and is not an empty directory.
src $ git clone https://github.com/sclorg/nodejs-ex
Cloning into 'nodejs-ex'...
remote: Enumerating objects: 631, done.
remote: Total 631 (delta 0), reused 0 (delta 0), pack-reused 631
Receiving objects: 100% (631/631), 235.11 KiB | 465.00 KiB/s, done.
Resolving deltas: 100% (243/243), done.
src $ cd nodejs-ex
nodejs-ex - master $ odo create nodejs
 ✓  Validating component [12ms]

Please use `odo push` command to create the component with source deployed
nodejs-ex - master! $ odo push
Validation
 ✓  Checking component [9ms]

Configuration changes
 ✓  Initializing component
 ✓  Creating component [70ms]

Pushing to component nodejs-nodejs-ex-awkc of type local
 ✓  Checking files for pushing [771831ns]
 ✓  Waiting for component to start [7s]
 ✓  Syncing files to the component [326ms]
 ✓  Building component [12s]
 ✓  Changes successfully pushed to component
nodejs-ex - master! $ odo describe
Component Name: nodejs-nodejs-ex-awkc
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858
nodejs-ex - master! $ odo url create
 ✓  URL nodejs-nodejs-ex-awkc-8080 created for component: nodejs-nodejs-ex-awkc

To create URL on the OpenShift Cluster, please use `odo push`
nodejs-ex - master! $ odo push
Validation
 ✓  Checking component [21ms]

Configuration changes
 ✓  Retrieving component data [25ms]
 ✓  Applying configuration [43ms]

Applying URL changes
 ✓  URL nodejs-nodejs-ex-awkc-8080: http://nodejs-nodejs-ex-awkc-8080-app-nphmecdnzx.192.168.42.122.nip.io created

Pushing to component nodejs-nodejs-ex-awkc of type local
 ✓  Checking file changes for pushing [968672ns]
 ✓  No file changes detected, skipping build. Use the '-f' flag to force the build.
nodejs-ex - master! $ odo describe
Component Name: nodejs-nodejs-ex-awkc
Type: nodejs
Source: file://./
Environment Variables:
 · DEBUG_PORT=5858
panic: runtime error: slice bounds out of range [:-1]

goroutine 1 [running]:
github.com/openshift/odo/pkg/odo/util.PrintComponentInfo(0xc00073c5a0, 0xc0008808c0, 0x15, 0x1925c0b, 0x9, 0x193a6df, 0x19, 0xc0008808c0, 0x15, 0x0, ...)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/util/cmdutils.go:142 +0x13b2
github.com/openshift/odo/pkg/odo/cli/component.(*DescribeOptions).Run(0xc00071fe90, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/describe.go:89 +0x2e0
github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun(0x1bc68a0, 0xc00071fe90, 0xc000887d40, 0x2a73ca8, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/genericclioptions/runnable.go:28 +0x100
github.com/openshift/odo/pkg/odo/cli/component.NewCmdDescribe.func1(0xc000887d40, 0x2a73ca8, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/pkg/odo/cli/component/describe.go:107 +0x5e
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).execute(0xc000887d40, 0x2a73ca8, 0x0, 0x0, 0xc000887d40, 0x2a73ca8)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:702 +0x285
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc00090a6c0, 0x0, 0x0, 0x0)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:783 +0x2c9
github.com/openshift/odo/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/home/dshah/go/src/github.com/openshift/odo/vendor/github.com/spf13/cobra/command.go:736
main.main()
	/home/dshah/go/src/github.com/openshift/odo/cmd/odo/odo.go:80 +0x53b
```

**Comment by dharmit** [comment]
@mik-dass maybe adding more tests could help us catch few issues like those I've pointed in above comments?

**Comment by mik-dass** [comment]
> URL created for a pushed component doesn't show up in odo describe output. But when we delete the component without doing odo push, the URL shows up in odo describe output. Is this expected behaviour? Shouldn't the URL be described in the first case as well?


@dharmit  It's working for me.
```
[mrinaldas@localhost nodejs-ex]$ odo delete -f
This component has following urls that will be deleted with component
URL named example with host example-app-qtwvxgjcyu.b6ff.rh-idev.openshiftapps.com having protocol http at port 8080
 ✓  Deleting component nodejs-nodejs-ex-nvge [3s]
 ✓  Component nodejs-nodejs-ex-nvge from application app has been deleted
[mrinaldas@localhost nodejs-ex]$ odo describe
Component Name: nodejs-nodejs-ex-nvge
Type: nodejs
Source: ./
URLs:
 · URL named example will be exposed via 8080/TCP
```

> should Environment Variables be mentioned in odo describe output when the component is not pushed? If we fetch them from the running pod then no.

Yes those environment variables are being fetched from a running pod.

> Output of Source is different for the two cases. Should we have consistency there?

I will have a look into how to make them consistent.

> @mik-dass maybe adding more tests could help us catch few issues like those I've pointed in above comments?

Currently we are missing proper checks for the `odo describe` output for pushed components. I have modified the tests and added some additional checks.

**Comment by mik-dass** [comment]
```
 ✗  Failed to create component with name ref-test. Please use `odo config view` to view settings used to create component. Error: unable to wait for build ref-test-app-1 to run: timed out waiting for the condition
[odo] unable get build log ref-test-app-1
[odo] github.com/openshift/odo/pkg/occlient.(*Client).FollowBuildLog
[odo] 	/home/travis/gopath/src/github.com/openshift/odo/pkg/occlient/occlient.go:1839
[odo] github.com/openshift/odo/pkg/component.Build
[odo] 	/home/travis/gopath/src/github.com/openshift/odo/pkg/component/component.go:794
[odo] github.com/openshift/odo/pkg/component.CreateComponent
[odo] 	/home/travis/gopath/src/github.com/openshift/odo/pkg/component/component.go:401
[odo] github.com/openshift/odo/pkg/odo/cli/component.(*CommonPushOptions).createCmpIfNotExistsAndApplyCmpConfig
[odo] 	/home/travis/gopath/src/github.com/openshift/odo/pkg/odo/cli/component/common_push.go:72
[odo] github.com/openshift/odo/pkg/odo/cli/component.(*CommonPushOptions).Push
[odo] 	/home/travis/gopath/src/github.com/openshift/odo/pkg/odo/cli/component/common_push.go:145
[odo] github.com/openshift/odo/pkg/odo/cli/component.(*PushOptions).Run
[odo] 	/home/travis/gopath/src/github.com/openshift/odo/pkg/odo/cli/component/push.go:105
[odo] github.com/openshift/odo/pkg/odo/genericclioptions.GenericRun
```
/retest

**Comment by mik-dass** [comment]
> Output of Source is different for the two cases. Should we have consistency there?

@dharmit Fixed

**Comment by mik-dass** [comment]
```
error: could not run steps: step integration-e2e-benchmark failed: template pod "integration-e2e-benchmark" failed: the pod ci-op-t78qx6bs/integration-e2e-benchmark failed after 1h49m33s (failed containers: lease): ContainerFailed one or more containers exited
```
/retest

File: tests/integration/component.go [hunk 1]
```diff
@@ -578,13 +596,21 @@ func componentTests(args ...string) {
...
 			// changing directory to the context directory
 			helper.Chdir(context)
 			cmpListOutput := helper.CmdShouldPass("odo", "list")
 			Expect(cmpListOutput).To(ContainSubstring(cmpName))
-			helper.CmdShouldPass("odo", "describe")
+			cmpDescribe := helper.CmdShouldPass("odo", "describe")
+
+			Expect(cmpDescribe).To(ContainSubstring(cmpName))
+			Expect(cmpDescribe).To(ContainSubstring("nodejs"))
+			Expect(cmpDescribe).To(ContainSubstring("file://./"))
```
Comment (amitkrout) [inline tests/integration/component.go]: won't work on windows platform. Use 
```
if runtime.GOOS == "windows" {
				sourcePath = "file:///./"
			} else {
				sourcePath = "file://./"
			}
```

File: tests/integration/component.go [hunk 2]
```diff
@@ -192,6 +192,24 @@ func componentTests(args ...string) {
 			helper.CmdShouldPass("odo", append(args, "delete", "-f", "--all", "--context", context)...)
 		})
 
+		It("should describe the component when it is not pushed", func() {
```
Comment (amitkrout) [inline tests/integration/component.go]: +1

**Comment by mik-dass** [comment]
@amitkrout Fixed

**Comment by amitkrout** [comment]
> @amitkrout Fixed

/lgtm

**Pull request merged**